### ISSUE TITLE
Rule Request: Remove redundant parenthesis from closures

### ISSUE THREAD

**Issue by rogerluan** [realm/SwiftLint#5252]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### New rule request

This new rule would remove redundant parenthesis from closures.

1. Why should this rule be added?

It's cleaner.

2. Provide several examples of what _would_ and _wouldn't_ trigger violations.

What would trigger:

```swift
let filteredResults = array.filter({ $0.isValid })
```

```swift
[1, 2].map({ $0 + 1 })
```

```swift
[1, 2].map({ $0 + 1 }).forEach({ print($0) })
```

```swift
[1, 2].map({ $0 + 1 })
  .forEach({ print($0) })
```

```swift
[1, 2].map { $0 + 1 }
  .map({ $0 + 2 })
  .map { $0 + 3 }
```

```swift
[1, 2].map({ $0 + 1 })
  .map { $0 + 2 }
  .map({ $0 + 3 })
```

What wouldn't trigger:

```swift
// Swift Syntax doesn't allow us to remove the parenthesis in this case
guard let filteredResults = array.filter({ $0.isValid }) else { return }
```

```swift
// Swift Syntax doesn't allow us to remove the parenthesis in this case
if let filteredResults = array.filter({ $0.isValid }) {
    print(filteredResults)
}
```

```swift
guard let validElement = array.first(where: { $0.isValid }) else { return }
```

```swift
let filteredResults = array.filter { $0.isValid }
```

```swift
[1, 2].map { $0 + 1 }
```

```swift
[1, 2].reduce(0) { $0 + $1 }
```

```swift
[1, 2].map { number in
 number + 1 
}
```

```swift
let isEmpty = [1, 2].isEmpty()
UIView.animateWithDuration(0.3, animations: {
   self.disableInteractionRightView.alpha = 0
}, completion: { _ in
   ()
})
```

Note that the last example need to play well alongside https://realm.github.io/SwiftLint/multiple_closures_with_trailing_closure.html

3. Should the rule be configurable, if so what parameters should be configurable?

I don't think there's any configuration needed.

4. Should the rule be opt-in or enabled by default? Why?

I think it should be enabled by default if, during implementation, we conclude that it's safe and doesn't have many false positives.

I **_believe_** there's a consensus that this is the expected way to write Swifty code, but I could be wrong. In fact I was very surprised that SwiftLint doesn't support this already. I'm curious to hear other opinions.

**Comment by marcelofabri** [comment]
Isn't this https://realm.github.io/SwiftLint/trailing_closure.html?

**Comment by rogerluan** [comment]
It is! 🤩 Thank you! I searched for rules named after "parenthesis" and "redundant", so I didn't find that one 😅 
Any clue why it's disabled by default?

Also I found out that SwiftFormat also has this rule: https://github.com/nicklockwood/SwiftFormat/blob/main/Rules.md#trailingclosures (in https://github.com/nicklockwood/SwiftFormat/issues/1537), and it even supports auto-correction so it's a win-win (if anyone's end up in this issue looking for such thing) 😊

**Comment by SimplyDanny** [comment]
> Any clue why it's disabled by default?

The rule might have a rather high rate of false positives. But as one can enable and disable all rules at will, it doesn't really matter what the default is.
